src/main/java/net/wasdev/wlp/common/plugins/util/ServerFeatureUtil.java

```diff
@@ -0,0 +1,314 @@
+                        if ("featureManager".equals(child.getNodeName())) {
+                            if (result == null) {
+                                result = new HashSet<String>();
+                            }
+                            result.addAll(parseFeatureManagerNode(child));
+                        } else if ("include".equals(child.getNodeName())){
+                            result = parseIncludeNode(result, canonicalServerFile, child, updatedParsedXmls);
+                        }
+                    }
+                }
+            } catch (IOException | ParserConfigurationException | SAXException e) {
+                // just skip this server.xml if it cannot be parsed
+                warn("The server file " + serverFile + " cannot be parsed. Skipping its features.");
+                debug(e);
+                return result;
+            }
+        }
+        return result;
+    }
+    
+    /**
+     * Parse feature elements from a featureManager node, trimming whitespace
+     * and treating everything as lowercase.
+     * 
+     * @param node
+     *            The featureManager node
+     * @return Set of trimmed lowercase feature names
+     */
+    private static Set<String> parseFeatureManagerNode(Element node) {
+        Set<String> result = new HashSet<String>();
+        NodeList features = node.getElementsByTagName("feature");
+        if (features != null) {
+            for (int j = 0; j < features.getLength(); j++) {
+                String content = features.item(j).getTextContent();
+                if (content != null) {
+                    if (content.contains(":")) {
+                        String[] split = content.split(":", 2);
+                        result.add(split[1].trim().toLowerCase());
+                    } else {
+                        result.add(content.trim().toLowerCase());
+                    }
+                }
+            }
+        }
+        return result;
+    }
+    
+    /**
+     * Parse features from an include node.
+     * 
+     * @param origResult
+     *            The features that have been parsed so far.
+     * @param serverFile
+     *            The parent server XML file containing the include node.
+     * @param node
+     *            The include node.
+     * @param updatedParsedXmls
+     *            The list of XML files that have been parsed so far.
+     * @return The set of features to install, or empty set if the cumulatively
+     *         parsed xml files only have featureManager sections but no
+     *         features to install, or null if there are no valid xml files or
+     *         they have no featureManager section
+     */
+    private Set<String> parseIncludeNode(Set<String> origResult, File serverFile, Element node,
+            List<File> updatedParsedXmls) {
+        Set<String> result = origResult;
+        String includeFileName = node.getAttribute("location");
```

Comment:
If the location attribute contains a reference to a variable, what happens here? Is that allowed? Not sure.

Reply:
This class was just refactored from the existing InstallFeatureUtil.  Regardless, the location attribute is in server.xml and it cannot have a reference to a variable because it needs to be able to work outside of Maven (e.g. it should work with the installUtility tool if using WebSphere Liberty)

Reply:
I just verified that the location attribute can have a variable in it (not a maven variable..that is not what I meant....a variable defined in a Liberty file (like server.env, another include file or server.xml) or an environment variable. https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_setup_includes.html

Reply:
Opened a separate issue to track this as it affects install feature as well.  https://github.com/WASdev/ci.common/issues/53